### DISCUSSION TITLE
Additional fields on `BaseUser`?

### DISCUSSION THREAD

**Discussion by dantownsend** [piccolo-orm/piccolo#512] [Ideas]
Piccolo has a [BaseUser](https://github.com/piccolo-orm/piccolo/blob/a89c2fa57ba01f352e6830b99ff108b50b14c0e8/piccolo/apps/user/tables.py#L21) table, which is used for authentication.

These are the columns it currently has:

- id
- username
- password
- first_name
- last_name
- email 
- active
- admin
- superuser
- last_login

A common pattern is to have a `Profile` table with a foreign key to `BaseUser` for storing additional information.

Should we add any additional columns to `BaseUser`?

These are the ones I'm considering:

- phone_number (for 2 factor)
- email_confirmed (has the user confirmed their email address)
- created_on
- title (Mr / Mrs etc)

There are others as well, like:

- address
- country
- timezone

**Comment by sinisaos** [comment]
@dantownsend I think we should keep the `BaseUser` as small and simple as possible (current fields are good enough). `BaseUser` can be [extended](https://piccolo-orm.readthedocs.io/en/latest/piccolo/authentication/baseuser.html#extending-baseuser) (changed in two ways) if someone need extra fields, as you said before.

**Reply by dantownsend** [reply]
Yeah, it's a good point - don't want to make it overly complex.

One thing I definitely regret not adding is `created_on` - that was an oversight really.

**Reply by sinisaos** [reply]
Good point. You can add a `created_on` column so we know when the user created account.

**Comment by dantownsend** [comment]
I was thinking about this today. One thing we could potentially do is add a `JSON` column called `extra`. That way any extra info can be stored in that field.

I'd like to add 2 factor support to Piccolo Admin at some point, and rather than adding an extra table or columns, we could just use `extra`, and support many different 2 factor providers.

**Comment by shahensaifullah** [comment]
how can i create fresh User model because sometimes it needs to make custom User model like my project is mainly login with phone number and otp. so here i don't need username. please make a baseuser where i can define username field like:
`username = 'phone_number'
`
this will help to set the username field.
